internal/controllers/pool/controller.go

```diff
@@ -0,0 +1,210 @@
+
+// Pool provides access to StorageOS Pools.
+//go:generate mockgen -build_flags=--mod=vendor -source=controller.go -destination=mocks/mock_pool.go -package=mocks . Pool
+type Pool interface {
+	GetPool(ctx context.Context, key client.ObjectKey) (*stosv1.Pool, error)
+	ListPools(ctx context.Context) (*stosv1.PoolList, error)
+	DeletePool(ctx context.Context, key client.ObjectKey) error
+	UpdatePool(ctx context.Context, pool *stosv1.Pool) (string, error)
+	CreatePool(ctx context.Context, pool *stosv1.Pool) (string, error)
+}
+
+// Reconciler reconciles a Pool object
+type Reconciler struct {
+	client.Client
+	log      logr.Logger
+	api      Pool
+	recorder record.EventRecorder
+}
+
+// NewReconciler returns a new SharedVolumeAPIReconciler.
+func NewReconciler(
+	api Pool,
+	k8s client.Client,
+	recorder record.EventRecorder) *Reconciler {
+	return &Reconciler{
+		Client:   k8s,
+		log:      ctrl.Log.WithName("controllers").WithName("Pool"),
+		api:      api,
+		recorder: recorder,
+	}
+}
+
+//+kubebuilder:rbac:groups=api.storageos.com,resources=pools,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=api.storageos.com,resources=pools/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=api.storageos.com,resources=pools/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	pool := &stosv1.Pool{}
+	if err := r.Client.Get(ctx, req.NamespacedName, pool); err != nil {
+		if errors.IsNotFound(err) {
+			// Pool not found (has been deleted on k8s by user)
+			// Call C2 API to delete pool on C2 if it exists
+			if err := r.api.DeletePool(ctx, req.NamespacedName); err != nil {
+				return ctrl.Result{}, err
+			}
+			// Delete related SC on k8s.
+			if err := r.deleteStorageClass(ctx, scName(scPrefix, req.Namespace, req.Name)); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			return ctrl.Result{}, nil
+		}
+		// Error getting object, requeue the request
+		return ctrl.Result{}, err
+	}
+
+	// Pool exists on k8s, call C2 API to check if pool exists on C2
+	_, err := r.api.GetPool(ctx, req.NamespacedName)
+	if err != nil {
+		if err == storageos.ErrPoolNotFound {
+			// Pool does not exist on C2, create it
+			pID, err := r.api.CreatePool(ctx, pool)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			// Create related storage class (update if it already exists)
+			if err := r.updateOrCreateStorageClass(ctx, newStorageClass(scPrefix, *pool, pID)); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// TODO: (CTRL-291)
+			// - Add node ID to pool status
+			// - Check drives listed in NodeDriveMap, update status.DrivesList with
+			//   drives that the pool consists of.
+
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Pool exists on C2, update it
```

Comment:
We just called the GetPool so we could check if we need to update it instead of just sending the request through everytime

Reply:
Another thing, the UpdatePool is internally doing another GetPool? If possible it would be better to only do a single request to the API here
Maybe create a different UpdatePool that takes the object (struct from the api) when you have it already, which is this case

Reply:
Both should be fixed now. 

UpdatePool no longer calls `GetPool` and we no longer update things unless we need to